episodes/5-transfer-learning.Rmd

```diff
@@ -0,0 +1,235 @@
+# input tensor
+inputs = keras.Input(train_images.shape[1:])
+```
+
+Our images are 64 x 64 pixels, whereas the pre-trained model that we will use was 
+trained on images of 160 x 160 pixels. To deal with this, we add an upscale layer
+that resizes the images to 160 x 160 pixels during training and prediction.
+```python
+# upscale layer
+method = tf.image.ResizeMethod.BILINEAR
+upscale = keras.layers.Lambda(
+  lambda x: tf.image.resize_with_pad(x, 160, 160, method=method))(inputs)
+```
+
+From the `keras.applications` module we use the `DenseNet121` architecture. 
+This architecture was proposed by the paper: [Densely Connected Convolutional Networks (CVPR 2017)](https://arxiv.org/abs/1608.06993). It is trained on the [Imagenet](https://www.image-net.org/) dataset, which contains 14,197,122 annotated images according to the WordNet hierarchy with over 20,000 classes.
+
+We will have a look at the architecture later, for now it is enough to know
+that it is a convolutional neural network with 121 layers that was designed 
+to work well on image classification tasks.
+
+Let's configure the DenseNet121:
+```python
+base_model = keras.applications.DenseNet121(include_top=False,
+                                            pooling='max'
+                                            weights='imagenet',
+                                            input_tensor=upscale,
+                                            input_shape=(160,160,3),
+                                            )
+```
+By setting `include_top` to `False` we exclude the fully connected layer at the
+top of the network. This layer was used to predict the Imagenet classes,
+but will be of no use for our dollarstreet dataset.
+
+We add `pooling='max'` so that max pooling is applied to the output of the DenseNet121 network.
+
+By setting `weights='imagenet'` we use the weights that resulted from training
+this network on the Imagenet data.
+
+We connect the network to the `upscale` layer that we defined before.
+
+### Only train a 'head' network
+Instead of fine-tuning all the weights of the DenseNet121 network using our dataset,
+we choose to freeze all these weights and only train a so-called 'head network' 
+that sits on top of the pre-trained network. You can see the DenseNet121 network
+as extracting a meaningful feature representation from our image. The head network
+will then be trained to decide on which of the 10 dollarstreet classes the image belongs.
+
+We will turn of the `trainable` property of the base model:
+```python
+base_model.trainable = False
+```
+
+Let's define our 'head' network:
+```python
+out = base_model.output
+out = keras.layers.Flatten()(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(128, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(64, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.Dense(10)(out)
```

Comment:
Not really important, but the same hyper-parameters (layers, layer sizes, dropout rates) as in the previous section for consistency might prevent confusion here.

Reply:
Good point, also it is overly complex.

I now use:
```python
out = base_model.output
out = keras.layers.Flatten()(out)
out = keras.layers.BatchNormalization()(out)
out = keras.layers.Dense(50, activation='relu')(out)
out = keras.layers.Dropout(0.5)(out)
out = keras.layers.Dense(10)(out)
```